Feature/ahans/pre commit clang format

ahans opened the pull request:
This is a draft of how the formatting could look like.

ahans commented on .clang-format:
```diff
@@ -0,0 +1,307 @@
+---
```
This is a dump of configuration, so I could play with all options. But I think this is largely based on some pre-defined style, so this can be drastically simplified, which I want to do before merging this.

keiravillekode replied:
We could delegate many of the style decisions by using [BasedOnStyle](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#basedonstyle)

When I worked on Chromium there was a short [.clang-format](https://github.com/chromium/chromium/blob/main/.clang-format)
The rendering engine originally used the `WebKit` style.

I don't have a preference between these or others.

ahans replied:
Yes, this is exactly what I was referring to when saying "simplify based on some pre-defined style".

ahans commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,15 @@
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
```
Those are some default hooks that are good to have IMHO. They're not required to run clang-format, though.

ahans commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,15 @@
+exclude: (.*/vendor/.*)
```
This excludes any file in a `vendor` directory. I assume those are copied from elsewhere and should probably not be touched by us.

ahans commented:
What's still missing from this is the integration into the GH workflow. For now, you can run this locally by installing [pre-commit](https://pre-commit.com/) and then doing `pre-commit run --all`. A CI job would run it also, but not apply anything automatically. Instead, the job will fail if `pre-commit run --all` would change anything, meaning that some fixes were not applied locally.

ahans commented on .github/CODEOWNERS:
```diff
@@ -3,4 +3,3 @@
 
 # Changes to `fetch-configlet` should be made in the `exercism/configlet` repo
 bin/fetch-configlet     @exercism/maintainers-admin
-
```
If we don't want to bug `@exercism/maintainers-admin` for this whitespace change, we can also add it to `exclude`.

keiravillekode replied:
We can add `bin/fetch-configlet` to the exclusions in `.pre-commit-config.yaml`

ahans replied:
The issue is not `bin/fetch-configlet`, but the extra blank line in `CODEOWNERS`. Honestly, I would also want to have that clean, but probably makes sense to do a follow-up PR where we only fix that, so that anyone from `@exercism/maintainers-admin` has to only review that bit and not all the rest.

keiravillekode replied:
We can add both `bin/fetch-configlet` and `CODEOWNERS` to the exclusions. Then this PR won't need review by `maintainers-admin`

keiravillekode commented:
```
pascals_triangle_test.c
```

It would be excellent if we could keep the triangular layout in the test expectations.

```
minesweeper_test.c
diamond_test.c
rectangles_test.c
state_of_tic_tac_toe_test.c
```

It would be excellent if we could keep the rectangular layout in the test inputs.

### Option A
The generator could output
```
// clang-format off
// clang-format on
```

### Option B
Perhaps the generator could output `//` at the end of each line in the array initializers to prevent the layout from being changed.

Lua test generators use an equivalent approach: 
[state-of-tic-tac-toe_spec.lua](https://github.com/exercism/lua/blob/main/exercises/practice/state-of-tic-tac-toe/state-of-tic-tac-toe_spec.lua)
[spec_generator.lua](https://github.com/exercism/lua/blob/main/exercises/practice/state-of-tic-tac-toe/.meta/spec_generator.lua)

ahans commented:
> ```
> pascals_triangle_test.c
> ```
> 
> It would be excellent if we could keep the triangular layout in the test expectations.
> 
> ```
> minesweeper_test.c
> diamond_test.c
> rectangles_test.c
> state_of_tic_tac_toe_test.c
> ```
> 
> It would be excellent if we could keep the rectangular layout in the test inputs/expectations.
> 
> ### Option A
> The generator could output
> 
> ```
> // clang-format off
> // clang-format on
> ```
> 
> ### Option B
> Perhaps the generator could output `//` at the end of each line in the array initializers to prevent the layout from being changed.
> 
> Lua test generators use an equivalent approach: [state-of-tic-tac-toe_spec.lua](https://github.com/exercism/lua/blob/main/exercises/practice/state-of-tic-tac-toe/state-of-tic-tac-toe_spec.lua) [spec_generator.lua](https://github.com/exercism/lua/blob/main/exercises/practice/state-of-tic-tac-toe/.meta/spec_generator.lua)

I quickly put that up yesterday to show that I'm working on something. I hadn't really looked at all the changes. I totally agree that the formatting of those test cases should not be touched. In general, when we only care about where to put line-breaks, having `//` at the end of the line is the way to go. However, for `pascals_triangle_test.c` that doesn't work, because there lines get too long as well. clang-format would break those even with a trailing `//`. So there we need the `// clang-format off`. 

Also, I totally forgot about the generators. Whatever formatting can be fixed there, we should definitely fix there. Then clang-format would only be a safety net.

I have this done for pascals_triangle already and will push that in a moment.

ahans commented on generators/exercises/pascals_triangle.py:
```diff
@@ -7,7 +7,7 @@
 #define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
 
-extern size_t rows(uint64_t* dest, size_t count);
+extern size_t rows(uint64_t *dest, size_t count);
```
This generator now creates a file that clang-format would not want to change.

keiravillekode commented:
Any progress or difficulties? Anything I can help with?

ahans commented:
> Any progress or difficulties? Anything I can help with?

No, I just haven't gotten around to continue here. Sorry for this! It's amazing how you keep adding exercises! After those PRs are merged this track has more than x86_64 now I think ...

ahans commented:
@keiravillekode  I think I addressed all open issues. Please have another look and sorry for the wait!

About minesweeper, to me it looks like clang-format does not change the rectangular layout, so no special handling is necessary here. Or am I missing something?

In general, I tried to adjust the generators so that they format the code correctly. However, line breaks of long strings or long arrays are a bit tricky and probably not worth implementing that in a generator. So I added the `--format` option to the `generate` script to have it run clang-format afterwards. This allows the `generate` script to leave us with test files that don't  need extra formatting without having to re-implement some of clang-format's logic in the generators.

ahans commented:
If you're fine with this, please don't just merge, but let me clean up the Git history first. It could make sense to do this with two or three separate commits, but currently it's obviously a mess. Thanks!

keiravillekode commented:
I often use "Squash and merge" and delete the generated commit summary.
